pkg/controller/nodeport-proxy/envoymanager/resources.go

```diff
@@ -27,21 +27,259 @@ import (
+	//	serviceKey := ServiceKey(service)
+	//	for _, servicePort := range service.Spec.Ports {
+	//		serviceNodePortName := fmt.Sprintf("%s-%d", serviceKey, servicePort.NodePort)
+
+	// ServiceKey returns a string: "Service.Namespace/Service.Name" so the whole string is:
+	// "namespace/name-nodeport"
+	// serviceNESName := fmt.Sprintf("%s-%d", serviceKey, service.Spec.Ports[0].Port)
+	serviceKey := ServiceKey(service)
+
+	tcpProxyConfig := &envoytcpfilterv3.TcpProxy{
+		StatPrefix: "ingress_tcp",
+		ClusterSpecifier: &envoytcpfilterv3.TcpProxy_Cluster{
+			Cluster: serviceKey,
+		},
+		AccessLog: makeAccessLog(),
+	}
+
+	tcpProxyConfigMarshalled, err := ptypes.MarshalAny(tcpProxyConfig)
+	if err != nil {
+		panic(errors.Wrap(err, "failed to marshal tcpProxyConfig"))
+	}
+
+	r.log.Debugf("Using a listener on port %d", HTTPSAltPort)
+
+	sniListener := &envoylistenerv3.Listener{
+		Name: serviceKey,
+		Address: &envoycorev3.Address{
+			Address: &envoycorev3.Address_SocketAddress{
+				SocketAddress: &envoycorev3.SocketAddress{
+					Protocol: envoycorev3.SocketAddress_TCP,
+					Address:  "0.0.0.0",
+					PortSpecifier: &envoycorev3.SocketAddress_PortValue{
+						PortValue: HTTPSAltPort,
+					},
+				},
+			},
+		},
+		FilterChains: []*envoylistenerv3.FilterChain{
+			{
+				Filters: []*envoylistenerv3.Filter{
+					{
+						Name: envoywellknown.TCPProxy,
+						ConfigType: &envoylistenerv3.Filter_TypedConfig{
+							TypedConfig: tcpProxyConfigMarshalled,
+						},
+					},
+				},
+				FilterChainMatch: makeSNIFilterChainMatch(service),
+			},
+		},
+	}
+	return sniListener
+}
+
+func makeTunnelingVirtualHosts(service *corev1.Service) []*envoyroutev3.VirtualHost {
+	serviceKey := ServiceKey(service)
+	virtualhosts := []*envoyroutev3.VirtualHost{
+		{
+			Name: serviceKey,
+			Domains: []string{
+				fmt.Sprintf("%s.%s.svc.cluster.local", service.Name, service.Namespace),
+			},
+			Routes: []*envoyroutev3.Route{
+				{
+					Match: &envoyroutev3.RouteMatch{
+						PathSpecifier: &envoyroutev3.RouteMatch_ConnectMatcher_{
+							ConnectMatcher: &envoyroutev3.RouteMatch_ConnectMatcher{},
+						},
+					},
+					Action: &envoyroutev3.Route_Route{
+						Route: &envoyroutev3.RouteAction{
+							ClusterSpecifier: &envoyroutev3.RouteAction_Cluster{
+								Cluster: serviceKey,
+							},
+							UpgradeConfigs: []*envoyroutev3.RouteAction_UpgradeConfig{
+								{
+									UpgradeType:   UpgradeType,
+									ConnectConfig: &envoyroutev3.RouteAction_UpgradeConfig_ConnectConfig{},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return virtualhosts
+}
+
+//TODO(youssefazrak) Currently we have a 1:1 bind between the listener itself
+// and the virtual_hosts. Each "tunnel" listener of the new expose strategy
+// should have two virtual_hosts, one for OpenVPN and one for the KAS.
+func (r *Reconciler) makeTunnelingListener(service *corev1.Service) *envoylistenerv3.Listener {
```

Comment:
I would add the tunneling and SNI listener to the initial resources, as they are immutable. We should add them only if the listener ports are given.